### PR TITLE
Redux

### PR DESCRIPTION
**Pull request by kaozdl**
this branch integrates redux with the app and allows the user to log in preserving the state.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#
+node_modules/
+npm-debug.log
+yarn-error.log
+
+# BUCK
+buck-out/
+\.buckd/
+*.keystore
+
```
Comment (nsantos16) [inline .gitignore]: Add a constant location 
```suggestion
# Constants location
.env
```

File: Api.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
+export default class ApiClient {
+  constructor() {
+    this.url = 'https://kaozdl-target.herokuapp.com/api/v1/';
```
Comment (nsantos16) [inline Api.js]: Add this to .env .

File: Api.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  }
+  async update(model, id, data) {
+    try {
+      let response = await fetch(this.url + model + '/' + id, {
+        method: 'PUT',
+        headers: {
+          Accept: 'appplication/json',
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(data),
```
Comment (nsantos16) [inline Api.js]: I recommend use axios library for requests([article](https://medium.com/@thejasonfile/fetch-vs-axios-js-for-making-http-requests-2b261cdd3af5)), but is okay if you stay this.

File: Api.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+      );
+      return response;
+    }
+    catch (error) {
+      return error;
+    }
+  }
+  async retrieve(model, id) {
+    try {
+      let response = await fetch(this.url + model + '/' + id).then(data);
```
Comment (nsantos16) [inline Api.js]: You can use string substitution on es6.
```suggestion
      let response = await fetch(`${this.url}model/${this.id}`).then(data);
```

Comment (kaozdl) [reply Api.js]: Thank's for the suggestion, that looks nicer

File: screens/signIn/styles.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    flex: 0.7,
+    justifyContent: 'center',
+    alignItems: 'center',
+  },
+  footer: {
+    flex: 0.3,
+    justifyContent: 'flex-start',
+    alignItems: 'center'
+  },
+  headerFont: {
```
Comment (nsantos16) [inline screens/signIn/styles.js]: Maybe a const file for fonts and styles in general.

File: screens/signIn/index.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+          <Button name={"SIGN IN"} action={this.handleSubmit}></Button>
+          <View style={{ padding: 20 }}>
+            <Text>Forgot your password?</Text>
+          </View>
+          <View style={{ padding: 20 }}>
+            <Text style={{ fontWeight: 'bold' }}>CONNECT WITH FACEBOOK</Text>
+          </View>
+        </View>
+        <View style={styles.footer}>
+          <Text>____________</Text>
```
Comment (nsantos16) [inline screens/signIn/index.js]: You can create a component for this, like Ruler.js, where you have to pass like a props the size of the ruler.

File: screens/signIn/index.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            bindTo={'username'}
+            handler={this.handleInput}
+          />
+          <Input
+            label={'PASSWORD'}
+            bindTo={'password'}
+            handler={this.handleInput}
+          />
+          <Button name={"SIGN IN"} action={this.handleSubmit}></Button>
+          <View style={{ padding: 20 }}>
```
Comment (nsantos16) [inline screens/signIn/index.js]: Add this to styles.js of this component.

File: screens/signIn/index.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+       })
+       .catch((response => {
+         alert("Login Error");
+       }));*/
+  }
+  render() {
+    return (
+      <Background>
+        <View style={styles.header}>
+          <Text style={styles.headerFont}>TARGET MVD</Text>
```
Comment (nsantos16) [inline screens/signIn/index.js]: Add the name of app to a const file.

File: screens/signIn/index.js [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  }
+  handleInput = (inputName, text) => this.setState({ [inputName]: text });
+  handleSubmit = () => {
+    const { username, password } = this.state;
+    const { signIn } = this.props;
+    console.log('PRELOGIN');
+    const user = { username: username, password: password };
+    this.props.login(username, password);
+    console.log('POSTLOGIN');
+    /* let response = await fetch('https://kaozdl-target.herokuapp.com/api/v1/rest-auth/login/', {
```
Comment (nsantos16) [inline screens/signIn/index.js]: Delete this comment

File: screens/signIn/index.js [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+      username: "",
+      password: "",
+    }
+  }
+  handleInput = (inputName, text) => this.setState({ [inputName]: text });
+  handleSubmit = () => {
+    const { username, password } = this.state;
+    const { signIn } = this.props;
+    console.log('PRELOGIN');
+    const user = { username: username, password: password };
```
Comment (nsantos16) [inline screens/signIn/index.js]: You can pass thought this json in this way.
```suggestion
    const user = { username , password };
```

File: screens/signIn/index.js [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    this.state = {
+      username: "",
+      password: "",
+    }
+  }
+  handleInput = (inputName, text) => this.setState({ [inputName]: text });
+  handleSubmit = () => {
+    const { username, password } = this.state;
+    const { signIn } = this.props;
+    console.log('PRELOGIN');
```
Comment (nsantos16) [inline screens/signIn/index.js]: Delete this.

File: screens/signIn/index.js [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+import { login } from '../../actions/session';
+import styles from './styles';
+import Background from '../common/background';
+
+import Input from '../../components/common/input';
+import Button from '../../components/common/button';
+
+
+class SignIn extends React.Component {
+  constructor() {
```
Comment (nsantos16) [inline screens/signIn/index.js]: You can delete this constructor and pass the state like this.
```suggestion
  state = {
      username: "",
      password: "",
    }
```

File: Api.js [hunk 13]
```diff
@@ -1,113 +1,35 @@
...
-        method: 'POST',
-        headers: {
-          Accept: 'application/json',
-          'Content-Type': 'application/json',
-        },
-        body: JSON.stringify(data),
+      body: JSON.stringify(data),
+    }).then(response => response.json()
+    ).then(response => resolve(response)
+    )
```
Comment (acollazomayer) [inline Api.js]: ).catch you have an extra enter

File: actions/session.js [hunk 14]
```diff
@@ -30,20 +29,16 @@ export const logout = () => async (dispatch) => {
...
-  catch (error) {
-    dispatch(loginError(error));
+    response = await apiLogin({ username, password });
+    console.log(response);
+    dispatch(loginSuccess(response.key));
+  } catch (error) {
+    const message = (error.status >= 500) ?
+      `Oops, something went wrong \n Error: ${error.status}` :
+      `Error: ${error.status} \n ${error.message}`
+    alert(message);
```
Comment (acollazomayer) [inline actions/session.js]: you should put this alert in the component, and the component is responsable to show the error.  Not the action. The error message should be dispatched in the action.

File: screens/signIn/index.js [hunk 15]
```diff
@@ -25,7 +25,14 @@ class SignIn extends React.Component {
   handleSignOut = () => { this.props.logout() };
   render() {
     let isLogged = this.props.session.isLoggedIn;
-    if (isLogged)
+    let isLoading = this.props.session.isLoading;
```
Comment (acollazomayer) [inline screens/signIn/index.js]: I know that you will refactor this in the next prs.

Comment (acollazomayer) [reply screens/signIn/index.js]: This screen into multiple screens

File: App.js [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+import { View } from 'react-native';
+import { Provider } from 'react-redux';
+import SignIn from './screens/signIn';
+import createStore from './store/configureStore';
+
+const store = createStore({});
+export default class App extends Component {
+  constructor(props) {
+    super(props);
+  }
```
Comment (juanlacruz) [inline App.js]: you don't need the constructor

File: reducers/session.js [hunk 17]
```diff
@@ -0,0 +1,26 @@
+import * as types from '../actions/actionTypes'
+
+const initialState = {
+  connected: false,
+  isLoading: false,
+  isLoggedIn: false,
+  userToken: '',
+}
+
+export default function targetApp(state = initialState, action) {
```
Comment (juanlacruz) [inline reducers/session.js]: later you could add inmutable.js to your store

Comment (kaozdl) [reply reducers/session.js]: I will study the library and add it soon.

File: App.js (hunk 16 above)
Comment (kaozdl) [reply App.js]: Noted, i will remove this.

**Pull request merged**